*** /usr/src/qmail-1.03/qmail-smtpd.c.bak	Mon Jun 15 05:53:16 1998
--- qmail-smtpd.c	Mon Jan  4 10:50:12 1999
*************** void die_ipme() { out("421 unable to fig
*** 50,55 ****
--- 50,56 ----
  void straynewline() { out("451 See http://pobox.com/~djb/docs/smtplf.html.\r\n"); flush(); _exit(1); }
  
  void err_bmf() { out("553 sorry, your envelope sender is in my badmailfrom list (#5.7.1)\r\n"); }
+ void err_bmt() { out("553 sorry, your envelope sender is in my badmailto list (#5.7.1)\r\n"); }
  void err_nogateway() { out("553 sorry, that domain isn't in my list of allowed rcpthosts (#5.7.1)\r\n"); }
  void err_unimpl() { out("502 unimplemented (#5.5.1)\r\n"); }
  void err_syntax() { out("555 syntax error (#5.5.4)\r\n"); }
*************** stralloc liphost = {0};
*** 96,101 ****
--- 97,105 ----
  int bmfok = 0;
  stralloc bmf = {0};
  struct constmap mapbmf;
+ int bmtok = 0;
+ stralloc bmt = {0};
+ struct constmap mapbmt;
  
  void setup()
  {
*************** void setup()
*** 117,122 ****
--- 121,131 ----
    if (bmfok)
      if (!constmap_init(&mapbmf,bmf.s,bmf.len,0)) die_nomem();
   
+   bmtok = control_readfile(&bmt,"control/badmailto",0);
+   if (bmtok == -1) die_control();
+   if (bmtok)
+     if (!constmap_init(&mapbmt,bmt.s,bmt.len,0)) die_nomem();
+  
    if (control_readint(&databytes,"control/databytes") == -1) die_control();
    x = env_get("DATABYTES");
    if (x) { scan_ulong(x,&u); databytes = u; }
*************** int bmfcheck()
*** 208,213 ****
--- 217,233 ----
    return 0;
  }
  
+ int bmtcheck()
+ {
+   int j;
+   if (!bmtok) return 0;
+   if (constmap(&mapbmt,addr.s,addr.len - 1)) return 1;
+   j = byte_rchr(addr.s,addr.len,'@');
+   if (j < addr.len)
+     if (constmap(&mapbmt,addr.s + j,addr.len - j - 1)) return 1;
+   return 0;
+ }
+ 
  int addrallowed()
  {
    int r;
*************** int addrallowed()
*** 219,224 ****
--- 239,245 ----
  
  int seenmail = 0;
  int flagbarf; /* defined if seenmail */
+ int bmtbarf; /* defined if seenmail */
  stralloc mailfrom = {0};
  stralloc rcptto = {0};
  
*************** void smtp_mail(arg) char *arg;
*** 250,256 ****
--- 271,279 ----
  void smtp_rcpt(arg) char *arg; {
    if (!seenmail) { err_wantmail(); return; }
    if (!addrparse(arg)) { err_syntax(); return; }
+   bmtbarf = bmtcheck();
    if (flagbarf) { err_bmf(); return; }
+   if (bmtbarf) { err_bmt(); return; }
    if (relayclient) {
      --addr.len;
      if (!stralloc_cats(&addr,relayclient)) die_nomem();
