--- TARGETS.orig  15-06-1998
+++ TARGETS       2010-55-08/05/10
--- ../qmail-1.03/TARGETS	1998-06-15 12:53:16.000000000 +0200
+++ ../qmail-1.03.071/TARGETS	2010-07-31 12:15:25.000000000 +0200
@@ -10,6 +10,7 @@
 qmail.o
 quote.o
 now.o
+base64.o
 gfrom.o
 myctime.o
 slurpclose.o
@@ -182,6 +183,8 @@
 qmail-remote
 qmail-rspawn.o
 tcpto_clean.o
+md5c.o
+hmac_md5.o
 qmail-rspawn
 direntry.h
 qmail-clean.o
